lib/saluki-components/src/transforms/name_filter/mod.rs

```diff
@@ -0,0 +1,238 @@
+
+        let i = self.data.binary_search(&name);
+
+        // Check for prefix match when match_prefix is true
+        if self.match_prefix {
+            if let Ok(index) = i {
+                if index > 0 && name.starts_with(&self.data[index - 1]) {
+                    return true;
+                }
+            } else if i.is_err() {
+                let index = i.unwrap_err();
+                if index > 0 && name.starts_with(&self.data[index - 1]) {
+                    return true;
+                }
+            }
+        }
+
+        if let Ok(index) = i {
+            return name == self.data[index];
+        }
+
+        false
+    }
+}
+
+#[derive(Debug)]
+pub struct NameFilter {
+    metric_prefix: String,
+
+    metric_prefix_blacklist: Vec<String>,
+
+    blocklist: Blocklist,
+}
+
+impl NameFilter {
+    fn enrich_metric(&self, mut metric: Metric) -> Option<Metric> {
+        let metric_name = metric.context().name().deref();
+        let mut new_metric_name = metric_name.to_string();
+
+        if !self.is_excluded(metric_name) {
+            new_metric_name = format!("{}{}", self.metric_prefix, metric_name);
+        }
+
+        if self.blocklist.test(new_metric_name.clone()) {
+            debug!("Metric {} excluded due to blocklist.", new_metric_name);
+            return None;
+        }
+
+        // Update metric with new name.
+        let new_context = metric.context().with_name(new_metric_name);
+        let existing_context = metric.context_mut();
+        *existing_context = new_context;
+
+        Some(metric)
+    }
+
+    fn is_excluded(&self, metric_name: &str) -> bool {
+        !self.metric_prefix.is_empty()
+            && self
+                .metric_prefix_blacklist
+                .iter()
+                .any(|prefix| metric_name.starts_with(prefix))
+    }
+}
+
+#[async_trait]
+impl Transform for NameFilter {
+    async fn run(mut self: Box<Self>, mut context: TransformContext) -> Result<(), GenericError> {
+        let mut health = context.take_health_handle();
+        health.mark_ready();
+
+        debug!("Agent metric name filter transform started.");
+
+        loop {
+            select! {
+                _ = health.live() => continue,
+                maybe_events = context.event_stream().next() => match maybe_events {
+                    Some(events) => {
+                        let mut buffered_forwarder = context.forwarder().buffered().expect("default output must always exist");
+
+                        for event in events {
+                            if let Some(metric) = event.try_into_metric() {
+                                if let Some(metric) = self.enrich_metric(metric) {
+
+                                let new_event = Metric::from_parts(
+                                    metric.context().clone(),
+                                    metric.values().clone(),
+                                    metric.metadata().clone(),
+                                );
```

Comment:
Why are we cloning/rebuilding the metric here? We just got it back from `enrich_metric`; it's already good to go.